miner/Miner.hs

```diff
@@ -0,0 +1,285 @@
+import Chainweb.Difficulty (HashTarget, encodeHashTarget)
+import Chainweb.RestAPI.Orphans ()
+import Chainweb.Time (Micros, Time, encodeTimeToWord64, getCurrentTimeIntegral)
+import Chainweb.Utils (int, runGet)
+import Chainweb.Version (ChainId, ChainwebVersion(..), chainwebVersionFromText)
+import Miner.Types ()
+
+---
+
+--------------------------------------------------------------------------------
+-- Servant
+
+type MinerAPI = "submit" :> ReqBody '[JSON] BlockHeader :> Post '[JSON] ()
+    :<|> "poll" :> Capture "chainid" ChainId
+                :> Capture "blockheight" BlockHeight
+                :> Get '[JSON] (Maybe BlockHeader)
+
+type API = MinerAPI :<|> SwaggerSchemaUI "help" "swagger.json"
+
+server :: Env -> Server API
+server e = (liftIO . submit e :<|> (\cid h -> liftIO $ poll e cid h))
+    :<|> swaggerSchemaUIServer (toSwagger (Proxy :: Proxy MinerAPI))
+
+app :: Env -> Application
+app = serve (Proxy :: Proxy API) . server
+
+--------------------------------------------------------------------------------
+-- CLI
+
+type ResultMap = HashMap (T2 ChainId BlockHeight) BlockHeader
+
+data Env = Env
+    { work :: TMVar BlockHeader
+    , results :: TVar ResultMap
+    , cores :: Word16
+    , version :: ChainwebVersion
+    , port :: Int
+    }
+
+pEnv :: TMVar BlockHeader -> TVar ResultMap -> Parser Env
+pEnv tbh thm = Env tbh thm <$> pCores <*> pVersion <*> pPort
+
+pCores :: Parser Word16
+pCores = option auto
+    (long "cores" <> metavar "COUNT" <> value 1
+     <> help "Number of CPU cores to use (default: 1)")
+
+pVersion :: Parser ChainwebVersion
+pVersion = option cver
+    (long "version" <> metavar "VERSION"
+     <> value Testnet01
+     <> help "Chainweb Network Version (default: testnet01)")
+  where
+    cver :: ReadM ChainwebVersion
+    cver = eitherReader $ \s ->
+        note "Illegal ChainwebVersion" . chainwebVersionFromText $ T.pack s
+
+pPort :: Parser Int
+pPort = option auto
+    (long "port" <> metavar "PORT" <> value 8081
+     <> help "Port on which to run the miner (default: 8081)")
+
+--------------------------------------------------------------------------------
+-- Work
+
+main :: IO ()
+main = do
+    env <- (opts <$> newEmptyTMVarIO <*> newTVarIO mempty) >>= execParser
+    miner <- async $ mining env
+    W.run (port env) $ app env
+    wait miner
+  where
+    opts :: TMVar BlockHeader -> TVar ResultMap -> ParserInfo Env
+    opts tbh thm = info (pEnv tbh thm <**> helper)
+        (fullDesc <> progDesc "The Official Chainweb Mining Client")
+
+-- | Submit a new `BlockHeader` to mine (i.e. to determine a valid `Nonce`).
+--
+submit :: Env -> BlockHeader -> IO ()
+submit (work -> w) bh = atomically $
+    isEmptyTMVar w >>= bool (void $ swapTMVar w bh) (putTMVar w bh)
+
+-- | For some `ChainId` and `BlockHeight`, have we mined a result?
+--
+poll :: Env -> ChainId -> BlockHeight -> IO (Maybe BlockHeader)
+poll (results -> tm) cid h = HM.lookup (T2 cid h) <$> readTVarIO tm
```

Comment:
Why/when is this API needed? How long is this information retained? What if we mined more than one block per chain and block height?

Reply:
> Why/when is this API needed?

Since the miner is push-based, some higher level coordinator (a `chainweb-node`, custom pool software, something else, etc.) must poll for the results of mining.

> How long is this information retained?

No more than 256 of the most recently mined `BlockHeader`s per chain are kept in a local in-memory cache.

> What if we mined more than one block per chain and block height?

This will only happen if the Cut-level coordinator (or pool) instructed the miner to mine on the same parent twice. If so, the newest result would overwrite the older.

Reply:
I don't think that this design will work well. Try to imagine how one would implement a client for that interface. That client would have to keep track of which blocks it submitted, it must keep state to remember to which cuts those belonged and what other blocks for other chains where submitted for that cut. Finally, it must loop in a tight loop to poll the results and push them to the cut pipeline, which requires a separate thread.

Reply:
Consider this example:

1. client pushes block candidate `a` for chain `x` at height `h` of cut `A` to miner.
2. miner starts working on `a`
3. client receives new cut and pushes block candidate `b` for chain `x` at height `h` of cut `B` to miner.
4. miner starts working on `b`.

Client poll for chain `x` at height `h` and receives a block. How does the client know for what cut the result is? It may be for cut `A` or cut `B`. The client could compare if the block matches the current cut (`B`) and reject the block if it doesn't fit. But what if the rejected block would actually provide a winning cut `A` that is better than `B`?

I think, all of this would be much simpler if the miner would be aware of cuts and would submit a new cut to the existing cut-pipeline API.

If the goal is that the mining logic, e.g. how to distribute hash power of different chains, should be implemented in the miner, the miner has to be aware of cuts in any case. So, a cut based API would be not only simpler but also more future proof.

Reply:
A cut aware miner could be implemented based on the existing Chainweb-consensus APIs. It could poll for new cuts on `GET cut` and push new cuts via `PUT cut`.

The only change in chainweb-consensus to support the new miner would be to allow to attach the block header to the cut, which is a feature that is already on our TODO, because it is needed for other things, too.

Reply:
> Finally, it must loop in a tight loop to poll the results and push them to the cut pipeline, which requires a separate thread.

You're right, and this is precisely how pool software works. The idea is that this "core miner" cares about "work" and nothing else. Consider this point, which tries to take a long view: it may not be a single `chainweb-node` instance that's curious about the mining results! In fact, it might not be a Haskell process written by us at all. With the current design, coordination between "Cut coordinators" and Miners is a potentially many-to-many relationship. If the Miner is instead made to "talk back" to some specific endpoint, that becomes one-to-one. 

Further, the fewer concerns a "core miner" has, the easier it will be to extend / reimplement. 

> How does the client know for what cut the result is?

This is solely an issue of how the miner handles its cache. We can make a decision now and be explicit about it in documentation. For instance, some rule like: if you submit work for chain `x` on height `h`, there will **never** be a result for it in the cache unless *that* specific job completed successfully. This implies that if the job for cut `A` had finished already, its result would be deleted from the cache when job `B` comes in for the same chain/height. Or we change the key type of the cache to be more granular.